pico/services/src/main/java/io/helidon/pico/services/DefaultServices.java

```diff
@@ -0,0 +1,481 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.helidon.pico.services;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+import io.helidon.pico.Application;
+import io.helidon.pico.CallingContext;
+import io.helidon.pico.CallingContextFactory;
+import io.helidon.pico.DefaultMetrics;
+import io.helidon.pico.DefaultQualifierAndValue;
+import io.helidon.pico.DefaultServiceInfoCriteria;
+import io.helidon.pico.InjectionException;
+import io.helidon.pico.Intercepted;
+import io.helidon.pico.Metrics;
+import io.helidon.pico.Module;
+import io.helidon.pico.Phase;
+import io.helidon.pico.PicoException;
+import io.helidon.pico.PicoServices;
+import io.helidon.pico.PicoServicesConfig;
+import io.helidon.pico.QualifierAndValue;
+import io.helidon.pico.Resettable;
+import io.helidon.pico.ServiceBinder;
+import io.helidon.pico.ServiceInfo;
+import io.helidon.pico.ServiceInfoCriteria;
+import io.helidon.pico.ServiceProvider;
+import io.helidon.pico.ServiceProviderBindable;
+import io.helidon.pico.ServiceProviderProvider;
+import io.helidon.pico.Services;
+
+import jakarta.inject.Provider;
+
+import static io.helidon.pico.CallingContext.toErrorMessage;
+
+/**
+ * The default reference implementation of {@link io.helidon.pico.Services}.
+ */
+class DefaultServices implements Services, ServiceBinder, Resettable {
+    private static final ServiceProviderComparator COMPARATOR = ServiceProviderComparator.create();
+
+    private final ConcurrentHashMap<String, ServiceProvider<?>> servicesByTypeName = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<String, Set<ServiceProvider<?>>> servicesByContract = new ConcurrentHashMap<>();
+    private final Map<ServiceInfoCriteria, List<ServiceProvider<?>>> cache = new ConcurrentHashMap<>();
+    private final PicoServicesConfig cfg;
+    private final AtomicInteger lookupCount = new AtomicInteger();
+    private final AtomicInteger cacheLookupCount = new AtomicInteger();
+    private final AtomicInteger cacheHitCount = new AtomicInteger();
+    private State stateWatchOnly; // we are watching and not mutating this state - owned by DefaultPicoServices
```

Comment:
If this is expected to be accessed from multiple threads (which it probably is), it should be declared volatile, or use `AtomicReference`

Reply:
pushed